src/ethereum/mod.rs

```diff
@@ -382,21 +406,36 @@ impl Ethereum {
         TX_COUNT.with_label_values(&[&bytes4]).inc();
 
         // Send TX to mempool
-        let pending = self
-            .provider
-            .send_transaction(tx.clone(), None)
-            .await
-            .map_err(|error| {
-                error!(?nonce, ?error, "Failed to send transaction");
-                TxError::Send(Box::new(error))
-            })?;
+        let span = span!(Level::INFO, "Send TX to mempool");
+        let span_guard = span.enter();
+        let pending = timeout(
+            self.send_timeout,
+            self.provider.send_transaction(tx.clone(), None),
+        )
+        .await
+        .map_err(|elapsed| {
+            error!(?elapsed, "Send transaction timed out");
+            TxError::SendTimeout
+        })?
+        .map_err(|error| {
+            error!(?nonce, ?error, "Failed to send transaction");
+            TxError::Send(Box::new(error))
+        })?;
+
         let tx_hash: H256 = *pending;
         info!(?nonce, ?tx_hash, "Transaction in mempool");
+        drop(span_guard);
 
         // Wait for TX to be mined
+        let span = span!(Level::INFO, "Wait for TX to be mined");
```

Comment:
I usually like to hide instrumentation as much as possible, here we're introducing multiple span instances within a single function and then we need to follow up with a drop to finish them I presume. Maybe we could introduce some functions to wrap actions  ?

Reply:
Yeah, devex could be improved. There is an alternative syntax:

```
info_span!("Wait for TX to be mined").in_scope(|| {
   // ... code in span
})
```

maybe I should use that more. (but also has some downsides, like hard to attach attributes to a span later). I need to experiment with this more.

Reply:
In most cases we can just add an `#[instrument]` atrtibute to the function. This one is an exception because there are a lot of complex steps happening that we want to log in detail. I could also split it up in multiple functions.